aptos-move/move-examples/fungible_asset/sources/fungible_source.move

```diff
@@ -0,0 +1,268 @@
+    use aptos_framework::object::generate_extend_ref;
+    #[test_only]
+    use std::signer::address_of;
+
+    /// The fungible asset supply exists or does not exist for this asset object.
+    const EFUNGIBLE_SOURCE: u64 = 1;
+    /// Amount cannot be zero.
+    const EZERO_AMOUNT: u64 = 2;
+    /// The current_supply of token as fungible assets is not zero.
+    const ECURRENT_SUPPLY_NON_ZERO: u64 = 3;
+    /// Mint capability exists or does not exist.
+    const EMINT_CAP: u64 = 4;
+    /// Freeze capability exists does not exist.
+    const EFREEZE_CAP: u64 = 5;
+    /// Burn capability exists or does not exist.
+    const EBURN_CAP: u64 = 6;
+    /// Current supply overflow
+    const ECURRENT_SUPPLY_OVERFLOW: u64 = 7;
+    /// Current supply underflow
+    const ECURRENT_SUPPLY_UNDERFLOW: u64 = 8;
+
+    #[resource_group_member(group = aptos_framework::object::ObjectGroup)]
+    struct FungibleSource has key {
+        current_supply: u64,
+        maximum_supply: Option<u64>
+    }
+
+    /// Capabilities
+    struct MintCap has store {
+        asset_addr: address
+    }
+
+    struct FreezeCap has store {
+        asset_addr: address
+    }
+
+    struct BurnCap has store {
+        asset_addr: address
+    }
+
+    public fun init_fungible_source(
+        extend_ref: &ExtendRef,
+        maximum_supply: u64
+    ): (MintCap, FreezeCap, BurnCap) {
+        assert_fungible_source_not_exists(address_from_extend_ref(extend_ref));
+        let asset_object_signer = object::generate_signer_for_extending(extend_ref);
+        let converted_maximum = if (maximum_supply == 0) {
+            option::none()
+        } else {
+            option::some(maximum_supply)
+        };
+        move_to(&asset_object_signer,
+            FungibleSource {
+                current_supply: 0,
+                maximum_supply: converted_maximum
+            }
+        );
+        let asset_addr = signer::address_of(&asset_object_signer);
+        (MintCap { asset_addr }, FreezeCap { asset_addr }, BurnCap { asset_addr })
+    }
+
+
+    public fun get_current_supply<T: key>(asset: &Object<T>): u64 acquires FungibleSource {
+        borrow_fungible_source(asset).current_supply
+    }
+
+    public fun get_maximum_supply<T: key>(asset: &Object<T>): Option<u64> acquires FungibleSource {
+        borrow_fungible_source(asset).maximum_supply
+    }
+
+    public fun assert_fungible_source_exists(asset_address: address) {
+        assert!(fungible_source_exists(asset_address), error::not_found(EFUNGIBLE_SOURCE));
+    }
+
+    public fun assert_fungible_source_not_exists(asset_address: address) {
+        assert!(!fungible_source_exists(asset_address), error::already_exists(EFUNGIBLE_SOURCE));
+    }
+
+    public fun fungible_source_exists(asset_address: address): bool {
+        exists<FungibleSource>(asset_address)
+    }
+
+    /// Mint the `amount` of coin with MintCap.
+    public fun mint_with_cap<T: key>(
+        cap: &MintCap,
+        asset: &Object<T>,
```

Comment:
the cap should sufficiently define the asset, no need for both or we have an issue.

Reply:
maybe we should call this mint_into?
how does one create a new object, can anyone do that? I need to read more.

Reply:
only the `create_signer` I hacked can create a new FA object... But you dislike it for no reason :(

Reply:
Option 1) Use the extendref that comes with the source, option 2) pass in the entity who will own the object.